### ISSUE TITLE
Correct amplitude of dft

### ISSUE THREAD

**Issue by lanougue** [xgcm/xrft#130]
Standard fft algorithm (numpy, ...) do not know coordinate spacing and then assumes it to be 1. The user have to manullay multiply the result by the spacing to get the true amplitude of the Fourier Transform.

For instance : In test_theoretical_matching(...), The transformed signal xrft.dft(s, dim='x') is multiplied by the spacing dx before comparison with theoretical result.

With xarray, we now know the coordinate spacing. I thus think that we should multiply the actual output of dft by the spacing to get the right Fourier Transform amplitude.

It would be a very simple but an important change in xrft. Moreover, it will make dft to be closer to what a true Fourier Transform is.

What are the opinion here ?

**Comment by roxyboy** [comment]
The theoretical FT and numpy FT end up having different units due to what you pointed out. I personally prefer to keep the `xrft.dft` consistent with the `numpy.fft` because the sum of variance computed by `numpy.fft` becomes equivalent to the sum of variance computed in the spatial dimension without having to care about the spatial `dx`.

**Comment by lanougue** [comment]
I agree with you but this is not exactly true.
You have to divide by the number of points (which is a kind of normalisation) to get the same number. Adding spacing to dft allow you to not care about the number of points and it looks like real parseval identity.

```
dx = 0.2
a = xr.DataArray(np.random.rand(500), dims='x', coords={'x':np.arange(500)*dx})
```

First case :
```
TFa = xrft.dft(a)
print(np.sum(np.abs(a.data)**2), np.sum(np.abs(TFa.data)**2)/len(TFa))
```

Second case :
```
TFa = xrft.dft(a)*dx
print(np.sum(np.abs(a.data)**2)*dx, np.sum(np.abs(TFa.data)**2)*TFa.freq_x_spacing)
```

**Comment by lanougue** [comment]
For xrft users, spacing in both direct and image domain are available. It should thus not be a problem.
Parsevel identity you mentioned in your comment remains true (if correctly computed as in the second case above).

I agree that it starts to depart from numpy.fft  but it is, for me, more logical since units would then be correct for both variances (direct and image domains).

**Comment by lanougue** [comment]
For zero mean signal `a-=a.mean()`, your identity would just writes:
```
a.var()*dx == TFa.var()*TFa.freq_x_spacing
```
Isn't it beautiful and more comparable to Parseval identity ? :)

**Comment by roxyboy** [comment]
In terms of computing the spectra, the convention (at least for me and papers I've seen in Oceanography) has been so that the `FTa` has the same dimension as `a` so, although I see your point, I'm still in favor of keeping it consistent with `numpy.fft`.

**Comment by lanougue** [comment]
@roxyboy, I do not really understand your point. Are you talking about units or dimension ? If it is dimensions, multiplying by the sampling will, of course, preserves the dimensions. If it is units, a Fourier Transform can not, by definition, have the same unit a the signal itself.

To use your own argument, multiplying fft output by the sampling allows one to not care of it in the analysis. For example, the exact same signal with two different sampling rate will give different results with fft algorithm. Adding the sampling in xrft.dft output would give the same result (which, I think, is the expected output of what a Fourier analysis is ideally meant for). Multiplying by dx not only provides the correct amplitude (and the right unit) but also allow the possibility to directly compare the frequency content of two different signals having two different sampling rate.

For a gobal point of view, I promote the fact to push xrft.dft the most close as possible to what a Fourier Transform is, instead of just mimicing fft algorihtm wich is, in fact, only a pure mathematical trick. FFT only do the numercial calculation, I beleive xrft.dft can do all what remains to do a complete Fourier Transform.

I see several advantages in this:
* No more fftshift or ifftshift stuff which is a problem the user had alway to deal with. This problem only comes from fft algorithm numerical implementation. With xarray and xrft the ffthsift problem totally desappears and becomes transparent for the user.
* With xarray, the sampling is known. I think, it would be sad not to use it to properly normalise the dft output. It directly provide the good normalisation associated with the corresponding unit.
* No need for the user to take care of the sampling to compare frequency content of signals with different sampling. 
* Number of points would remain a "numerical" parameters and would also become transparent for the user.
* In the inverse Fourier Transform, no need to remember if I need to multiply ou divise by the number of point. This will be again natural. The 2 pi will also be correctly added.
It is, in fact, a way to make the numerical considerations transparent for the user while ensuring the closest behaviour of the transformation to a Fourier Transform.

To summarize and to try to be constructive, I beleive that xrft/xarray have already all the means to do a complete Fourier Transform. FFT algorithm is, for me, only the numerical trick to compute it fastly. Removing the defaults of fft algorithm (shifts, wrong normalisation, sampling dependent, coordinate range not known) and taking full advantages of xarray seems the right way to proceed from my point of view.

Maybe, for a smooth update, it could be a possible option to provide two functions ? I remember that renaming dft into fft was an option. Perhaps, xrft could provide one function that mimic fft, and another wich provides the true Fourier Tansform ?

I would apreciate @rabernat and others opinion on that.

**Comment by roxyboy** [comment]
> Are you talking about units or dimension ? If it is dimensions, multiplying by the sampling will, of course, preserves the dimensions. If it is units, a Fourier Transform can not, by definition, have the same unit a the signal itself.

I'm not sure I agree with your last point. Based on how `numpy.fft` is implemented, I think the amplitude of Fourier Transform will have the same units as the signal itself.

**Comment by roxyboy** [comment]
@lanougue But I suppose we could have a flag like `theory=True` which would allow for using the metadata of spacing for normalization and a closer relation to the theory of Fourier transform.

**Comment by lanougue** [comment]
Yes, it can be done like this but, don't you think it can confuse the user with two much flags ? true_phase, true_amplitude, ...

**Comment by roxyboy** [comment]
I personally think it's better to preserve the functionality similar to `numpy.fft` but others can chime in.

**Comment by roxyboy** [comment]
The problem with changing the default functionality is that it will also confuse current users of `xrft` (which there aren't that many but still exist) and potentially break their pre-existing code. So, I'd recommend a soft change with a flag.

**Comment by lanougue** [comment]
I agree.
However, dft starts having numerous number of flags (for phase or amplitude). Don't you think a separated function could be a better option ?

**Comment by roxyboy** [comment]
Maybe we can have `xrft.fft` that calls `xrft.dft` with the flag of `true_amplitude`, i.e. `xrft.fft(..., true_amplitude=True)` without putting the flag in `dft` itself.

**Comment by roxyboy** [comment]
For `xrft.fft`, when `true_amplitude=False`, I think it should just return `xrft.dft`

**Comment by lanougue** [comment]
Ok, I think it is a very good option.
Following your spirit, and if I understand correctly, we could define xrft.fft that would mimic np.fft (and would thus call dft() with `true_amplitude=False` and `true_phase=False`). 
xrft.dft() will thus support the deprecation warning which says that true_phase and true_amplitude will be set to True by default in future dft() definition.
* xrft.fft() will thus return usual fft behaviour
* xrft.dft() will thus return true Fourier transform (by default only in future version)

Am I correct ?

**Comment by roxyboy** [comment]
Almost. What I had in mind was the opposite. We keep `xrft.dft` as is and add a [Deprecation Warning](https://docs.python.org/3/library/exceptions.html#PendingDeprecationWarning) to it that in future releases, it will be superceded by `xrft.fft`. The default `xrft.fft` behavior will be `xrft.fft(true_phase=True, true_amplitude=True)` and the two flags being false will return the current default `xrft.dft` behavior. In other words, I'm suggesting not to edit `xrft.dft` other than adding the warning.

**Comment by roxyboy** [comment]
Current users are calling `xrft.dft` with the expectation of `numpy.fft` functionality. For consistency, I think the default setting of `xrft.dft` should not be changed.

**Comment by lanougue** [comment]
In time, with your suggestion, the users will finally call xrft.fft() which will not mimic np.fft() anymore. I think this is strange.

With my previous suggestion, dft will continue to keep the same current fft behaviour up to an other new xrft release.
Meanwhile, with the warning, the user will have time to move from xrft.dft() to xrft.fft() that will actually mimic np.fft().
Only dft() will have true_flags turned to default True i nthe future.

At the end, xrft.fft() will match np.fft() and xrft.dft() will compute the real Fourier Transform. I think this is more logical on a long term perspective.

**Comment by roxyboy** [comment]
Makes sense. So we'll have `xrft.fft` do the heavy lifting and the default behavior of `xrft.dft` will be to call `xrft.fft` with the flags being false.

**Comment by roxyboy** [comment]
And we'll put a deprecation warning that in the future, the default behavior of `xrft.dft` will change to the flags being true. This is what you have in mind right @lanougue ?

**Comment by lanougue** [comment]
Almost.
dft() requires more computation steps (lag, phase, dx) that have to be computed prior to fft algorithm. 
Thus, I was thinking to make dft supports all the heavy code and the warning. dft would have default (amplitude/phase) flags to False in the next version and will be passed to True in the other next one.
xrft.fft() would simply call dft() with flags set to False.

This would simplify the xrft.fft() argument list compared to the dft one(). I think this is an easier way for new comers in xrft world

**Comment by roxyboy** [comment]
Your logic would mean that `xrft.fft` would always raise a warning in the next release but... I guess that's fine.

**Comment by lanougue** [comment]
Yes. Unless we find a way to deactivate the warning if dft is called through fft.

**Comment by lanougue** [comment]
If you are ok, I can go with that. @rabernat : what do you think ?

**Comment by roxyboy** [comment]
> Yes. Unless we find a way to deactivate the warning if dft is called through fft.

Seems like this would do the trick: https://docs.python.org/2/library/warnings.html#temporarily-suppressing-warnings

**Comment by lanougue** [comment]
Perfect ! I can start coding this if this plan suits.
Do I have to wait for @rabernat  ?

**Comment by roxyboy** [comment]
> dft() requires more computation steps (lag, phase, dx) that have to be computed prior to fft algorithm.
> Thus, I was thinking to make dft supports all the heavy code and the warning. dft would have default (amplitude/phase) flags to False in the next version and will be passed to True in the other next one.
> xrft.fft() would simply call dft() with flags set to False.
> 
> This would simplify the xrft.fft() argument list compared to the dft one(). I think this is an easier way for new comers in xrft world

He gave a thumbs up to your plan so I think you can go ahead.

**Comment by lanougue** [comment]
Perfect, I'll go for that then !
